src/real_ip.rs

```diff
@@ -0,0 +1,299 @@
+    "18.238.0.0/15",
+    "18.244.0.0/15",
+    "34.195.252.0/24",
+    "34.216.51.0/25",
+    "34.223.12.224/27",
+    "34.223.80.192/26",
+    "34.226.14.0/24",
+    "35.158.136.0/24",
+    "35.162.63.192/26",
+    "35.167.191.128/26",
+    "36.103.232.0/25",
+    "36.103.232.128/26",
+    "44.227.178.0/24",
+    "44.234.90.252/30",
+    "44.234.108.128/25",
+    "52.15.127.128/26",
+    "52.46.0.0/18",
+    "52.47.139.0/24",
+    "52.52.191.128/26",
+    "52.56.127.0/25",
+    "52.57.254.0/24",
+    "52.66.194.128/26",
+    "52.78.247.128/26",
+    "52.82.128.0/19",
+    "52.84.0.0/15",
+    "52.124.128.0/17",
+    "52.199.127.192/26",
+    "52.212.248.0/26",
+    "52.220.191.0/26",
+    "52.222.128.0/17",
+    "54.182.0.0/16",
+    "54.192.0.0/16",
+    "54.230.0.0/17",
+    "54.230.128.0/18",
+    "54.230.200.0/21",
+    "54.230.208.0/20",
+    "54.230.224.0/19",
+    "54.233.255.128/26",
+    "54.239.128.0/18",
+    "54.239.192.0/19",
+    "54.240.128.0/18",
+    "58.254.138.0/25",
+    "58.254.138.128/26",
+    "64.252.64.0/18",
+    "64.252.128.0/18",
+    "65.8.0.0/16",
+    "65.9.0.0/17",
+    "65.9.128.0/18",
+    "70.132.0.0/18",
+    "71.152.0.0/17",
+    "99.79.169.0/24",
+    "99.84.0.0/16",
+    "99.86.0.0/16",
+    "108.138.0.0/15",
+    "108.156.0.0/14",
+    "116.129.226.0/25",
+    "116.129.226.128/26",
+    "118.193.97.64/26",
+    "118.193.97.128/25",
+    "119.147.182.0/25",
+    "119.147.182.128/26",
+    "120.52.12.64/26",
+    "120.52.22.96/27",
+    "120.52.39.128/27",
+    "120.52.153.192/26",
+    "120.232.236.0/25",
+    "120.232.236.128/26",
+    "120.253.240.192/26",
+    "120.253.241.160/27",
+    "120.253.245.128/26",
+    "120.253.245.192/27",
+    "130.176.0.0/17",
+    "130.176.128.0/18",
+    "130.176.192.0/19",
+    "130.176.224.0/20",
+    "143.204.0.0/16",
+    "144.220.0.0/16",
+    "180.163.57.0/25",
+    "180.163.57.128/26",
+    "204.246.164.0/22",
+    "204.246.168.0/22",
+    "204.246.172.0/24",
+    "204.246.173.0/24",
+    "204.246.174.0/23",
+    "204.246.176.0/20",
+    "205.251.200.0/21",
+    "205.251.208.0/20",
+    "205.251.249.0/24",
+    "205.251.250.0/23",
+    "205.251.252.0/23",
+    "205.251.254.0/24",
+    "216.137.32.0/19",
+    "223.71.11.0/27",
+    "223.71.71.96/27",
+    "223.71.71.128/25",
+];
+
+static CLOUD_FRONT_NETWORKS: Lazy<Vec<IpNetwork>> = Lazy::new(|| {
+    CLOUD_FRONT_STRS
+        .iter()
+        .map(|s| s.parse().unwrap())
+        .collect()
+});
+
+fn is_cloud_front_ip(ip: &IpAddr) -> bool {
+    CLOUD_FRONT_NETWORKS
+        .iter()
+        .any(|trusted_proxy| trusted_proxy.contains(*ip))
+}
+
+pub fn process_xff_headers(headers: &HeaderMap) -> Option<IpAddr> {
+    let mut xff_iter = headers.get_all(X_FORWARDED_FOR).iter();
+    let first_header = xff_iter.next()?;
+
+    let has_more_headers = xff_iter.next().is_some();
+    return if has_more_headers {
+        // This only happens for requests going directly to crates-io.herokuapp.com,
```

Comment:
Are there cases where direct calls to `crates-io.herokuapp.com` should even actually be supported? As a DDOS protection method, wouldn't it make sense to *only* support calls directly to our utilized CDN proxies?

Reply:
> Are there cases where direct calls to `crates-io.herokuapp.com` should even actually be supported?

no

> As a DDOS protection method, wouldn't it make sense to _only_ support calls directly to our utilized CDN proxies?

yes

in other words: I fully agree with you, we just haven't gotten to that part yet. I don't think this is a blocker for this PR though and can be worked on in parallel :)